### DISCUSSION TITLE
Nuxt 3 - "No urql Client was provided. Did you forget to install the plugin or call provideClient in a parent?"

### DISCUSSION THREAD

**Discussion by michealroberts** [urql-graphql/urql#2122] [Q&A] [answered]
I have setup the following useURQ() composable:

```ts
import { createClient, provideClient, useClientHandle } from '@urql/vue'

import { dedupExchange, cacheExchange, fetchExchange, ssrExchange } from '@urql/core'

export interface UseURQLOptions {
  // URI representing a GraphQL endpoint:
  url: string
}

/**
 * useURQL()
 * Returns an URQL client instance based off of an input uri
 *
 * @param options of type UseURQLOptions
 * @output URQL client, useQuery and useMutation
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useURQL = (options: UseURQLOptions) => {
  const { url } = options

  const isServerSide = typeof window === 'undefined'

  // The `ssrExchange` must be initialized with `isClient` and `initialState`
  const ssr = ssrExchange({
    isClient: !isServerSide,
    initialState: undefined
  })

  const client = createClient({
    url: url,
    exchanges: [
      dedupExchange,
      cacheExchange,
      ssr,
      fetchExchange,
    ]
  })

  return {
    client,
    provideClient,
    useClientHandle
  }
}
```

And usage would be inside of a component's setup() method is as follows:

```
setup() {
    const { client, provideClient, useClientHandle } = useURQL({
      url: 'https://countries.trevorblades.com'
    })

    provideClient(client)

   ...

}
```

**However, I'm seeing this error:**

```
No urql Client was provided. Did you forget to install the plugin or call provideClient in a parent?
```

This is within a Nuxt 3 environment ... 

Could someone advise on the usage of this library within a composable please ...

**Comment by kitten** [comment]
Moving this to a discussion, since it's likely a usage issue;
That said, have you checked where you're using urql hooks? They can only be used below the component that provides the client

**Reply by michealroberts** [reply]
So let's say I have index.vue (Nuxt 3) root page ... this is where I effectively call "provideClient" ... I then can useQuery inside a child component of Index.vue ... ?

**Reply by michealroberts** [reply]
Looking at `defineNuxtPlugin` ... is there a tried and tested way to get this working with Nuxt 3 globally?

**Reply by michealroberts** [reply]
It might be worth noting that ApolloQL can be provided, and used without the need for parent provide / child injection ... however, it's also worth noting that this library is heads above ApolloQL (builds nicely with Nuxt 3). :D

**Comment by michealroberts** [comment, marked as answer]
Hoping this might help anyone working with urql on Nuxt 3:

```ts
import { defineNuxtPlugin } from '#app'

import urql from '@urql/vue'

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.vueApp.use(urql, {
    url: 'https://countries.trevorblades.com'
  })
})
```

**Reply by lewebsimple** [reply]
It also works by providing a ref to the client like so:
```
import { defineNuxtPlugin } from '#app'
import { ref } from 'vue';
import { createClient } from '@urql/vue';

export default defineNuxtPlugin((nuxtApp) => {
  const urqlClient = ref(createClient({ url: 'http://localhost:3000/api/graphql' }));
  nuxtApp.provide('urql', urqlClient);
  nuxtApp.vueApp.provide('$urql', urqlClient);
});
```

**Reply by lewebsimple** [reply]
For anybody getting this error in production:
```
Error: use* functions may only be called during the `setup()` or other lifecycle hooks.
```
Please remember to add `build:{ transpile:['@urql/vue'] }` inside your `nuxt.config.ts` !

**Reply by genu** [reply]
I'm seeing the `use*` error whenever I try to do any urql operations inside of a middleware. 

That should be possible, right?
